bindings/resolver.go

```diff
@@ -0,0 +1,253 @@
+		if err != nil {
+			return nil, err
+		}
+
+		var binding Binding
+		if isLegacy {
+			binding, err = loadLegacyBinding(bindingRoot, file.Name())
+		} else {
+			binding, err = loadBinding(bindingRoot, file.Name())
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "reading binding '%s'", file.Name())
+		}
+		bindings = append(bindings, binding)
+	}
+	return bindings, nil
+}
+
+// According to the legacy spec (https://github.com/buildpacks/spec/blob/main/extensions/bindings.md), a legacy binding
+// has a `metadata` directory within the binding path.
+func isLegacyBinding(bindingRoot string, name string) (bool, error) {
+	info, err := os.Stat(filepath.Join(bindingRoot, name, "metadata"))
+	if err == nil {
+		return info.IsDir(), nil
+	} else if os.IsNotExist(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+// See: https://github.com/k8s-service-bindings/spec#workload-projection
+func loadBinding(bindingRoot string, name string) (Binding, error) {
+	binding := Binding{
+		Name:   name,
+		Path:   filepath.Join(bindingRoot, name),
+		Secret: map[string]SecretReader{},
+	}
+
+	secrets, err := loadSecrets(filepath.Join(binding.Path))
+	if err != nil {
+		return Binding{}, err
+	}
+
+	typ, ok := secrets["type"]
+	if !ok {
+		return Binding{}, errors.New("missing 'type'")
+	}
+	binding.Type = typ.String()
+	delete(secrets, "type")
+
+	provider, ok := secrets["provider"]
+	if ok {
+		binding.Provider = provider.String()
+		delete(secrets, "provider")
+	}
+
+	binding.Secret = secrets
+
+	return binding, nil
+}
+
+// See: https://github.com/buildpacks/spec/blob/main/extensions/bindings.md
+func loadLegacyBinding(bindingRoot string, name string) (Binding, error) {
+	binding := Binding{
+		Name:   name,
+		Path:   filepath.Join(bindingRoot, name),
+		Secret: map[string]SecretReader{},
+	}
+
+	metadata, err := loadSecrets(filepath.Join(binding.Path, "metadata"))
+	if err != nil {
+		return Binding{}, err
+	}
+
+	typ, ok := metadata["kind"]
+	if !ok {
+		return Binding{}, errors.New("missing 'kind'")
+	}
+	binding.Type = typ.String()
+	delete(metadata, "kind")
+
+	provider, ok := metadata["provider"]
+	if !ok {
+		return Binding{}, errors.New("missing 'provider'")
+	}
+	binding.Provider = provider.String()
+	delete(metadata, "provider")
+
+	binding.Secret = metadata
+
+	secrets, err := loadSecrets(filepath.Join(binding.Path, "secret"))
+	if err != nil && !os.IsNotExist(err) {
+		return Binding{}, err
+	}
+	if err == nil {
+		for k, v := range secrets {
+			binding.Secret[k] = v
+		}
+	}
+
+	return binding, nil
+}
+
+func loadSecrets(secretsPath string) (map[string]SecretReader, error) {
+	secrets := map[string]SecretReader{}
+	files, err := ioutil.ReadDir(secretsPath)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, file := range files {
+		content, err := os.ReadFile(filepath.Join(secretsPath, file.Name()))
+		if err != nil {
+			return nil, err
+		}
+		secrets[file.Name()] = MakeSecretReader(content)
```

Comment:
I think I've come around to the idea that we should not read the whole contents into memory before the point at which someone calls `String()`, `Bytes()`, or `Read(b []byte)` on the `SecretReader`. Instead, could the implementation lazily read these contents when those methods are invoked? I think we'd also need to add `Close()` to the API contract to make this work well without exhausting file descriptors in some edgecases.

Basically, I am concerned about a case where a secret is a very large file, like a binary or tarball, that we then read the entire contents of into memory.

Reply:
That shouldn't be difficult to change. I'll look into it. I'll have to move away from using `bytes.Buffer` under the hood and may just use the file object

Reply:
Done